src/test/java/org/prebid/server/settings/service/S3PeriodicRefreshServiceTest.java

```diff
@@ -0,0 +1,206 @@
+    private static final String STORED_REQ_DIR = "stored-req";
+    private static final String STORED_IMP_DIR = "stored-imp";
+
+    @Mock
+    private CacheNotificationListener cacheNotificationListener;
+    @Mock
+    private Vertx vertx;
+    @Mock(strictness = LENIENT)
+    private S3AsyncClient s3AsyncClient;
+    private final Clock clock = Clock.fixed(Instant.now(), ZoneId.systemDefault());
+    @Mock
+    private Metrics metrics;
+
+    private final Map<String, String> expectedRequests = singletonMap("id1", "value1");
+    private final Map<String, String> expectedImps = singletonMap("id2", "value2");
+
+    @BeforeEach
+    public void setUp() {
+        given(s3AsyncClient.listObjects(any(ListObjectsRequest.class)))
+                .willReturn(listObjectResponse(STORED_REQ_DIR + "/id1.json"),
+                        listObjectResponse(STORED_IMP_DIR + "/id2.json"));
+
+        given(s3AsyncClient.getObject(any(GetObjectRequest.class), any(AsyncResponseTransformer.class)))
+                .willReturn(CompletableFuture.completedFuture(
+                                ResponseBytes.fromByteArray(
+                                        GetObjectResponse.builder().build(),
+                                        "value1".getBytes())),
+                        CompletableFuture.completedFuture(
+                                ResponseBytes.fromByteArray(
+                                        GetObjectResponse.builder().build(),
+                                        "value2".getBytes())));
+    }
+
+    @Test
+    public void shouldCallSaveWithExpectedParameters() {
+        // when
+        createAndInitService(1000);
+
+        // then
+        verify(cacheNotificationListener).save(expectedRequests, expectedImps);
+    }
+
+    @Test
+    public void shouldCallInvalidateAndSaveWithExpectedParameters() {
+        // given
+        given(vertx.setPeriodic(anyLong(), any()))
+                .willAnswer(withSelfAndPassObjectToHandler(1L));
+        given(s3AsyncClient.listObjects(any(ListObjectsRequest.class)))
+                .willReturn(listObjectResponse(STORED_REQ_DIR + "/id1.json"),
+                        listObjectResponse(STORED_IMP_DIR + "/id2.json"),
+                        listObjectResponse(),
+                        listObjectResponse(STORED_IMP_DIR + "/id2.json"));
+        given(s3AsyncClient.getObject(any(GetObjectRequest.class), any(AsyncResponseTransformer.class)))
+                .willReturn(getObjectResponse("value1"),
+                        getObjectResponse("value2"),
+                        getObjectResponse("changed_value"));
+
+        // when
+        createAndInitService(1000);
+
+        // then
+        verify(cacheNotificationListener).save(expectedRequests, expectedImps);
+        verify(cacheNotificationListener).invalidate(singletonList("id1"), emptyList());
+        verify(cacheNotificationListener).save(emptyMap(), singletonMap("id2", "changed_value"));
```

Comment:
This test case is now broken. My feeling is, that there's a race condition and the assertions are invoked, before the method is called. I'm trying to dig deeper on this

```
Wanted but not invoked:
cacheNotificationListener.invalidate(
    [id1],
    []
);
```

Reply:
Nope. This seems to be a problem in the business logic. Trying to figure out what it is